go2tv: init at 1.13.0

gdamjan opened the pull request:
Cast media files to UPnP/DLNA Media Renderers and Smart TVs.
https://github.com/alexballas/go2tv

provides 2 packages:
- go2tv the GUI version (based on go-gl/glfw)
- go2tv-lite just a cli version

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

SuperSandro2000 commented on pkgs/applications/video/go2tv/default.nix:
```diff
@@ -0,0 +1,41 @@
+  meta = with lib; {
+    description = "Cast media files to UPnP/DLNA Media Renderers and Smart TVs";
+    homepage = "https://github.com/alexballas/go2tv";
+    license = licenses.mit;
```
Missing meta.maintainers

gdamjan replied:
added myself

SuperSandro2000 commented on pkgs/applications/video/go2tv/default.nix:
```diff
@@ -0,0 +1,41 @@
+  doCheck = false;
+
+  buildInputs = [
+    xorg.libX11.dev
+    xorg.libXcursor.dev
+    xorg.libXrandr.dev
+    xorg.libXinerama.dev
+    xorg.libXi.dev
+    xorg.libXext.dev
```
dev output should be choosen automatically

gdamjan replied:
removed `.dev`

gdamjan commented:
updated to create 2 separate packages, one just cli (go2tv-lite), the other a gui version (based on go-gl/glfw).


I'm still not sure what's the proper way to provide the optional dependency of ffmpeg. go2tv just runs the executable if it's available, to do transcoding.

SuperSandro2000 commented:
> I'm still not sure what's the proper way to provide the optional dependency of ffmpeg. go2tv just runs the executable if it's available, to do transcoding.

If we don't want to add it to a wrapper all the time there is none. Optional wrappers just to populate PATH are overkill.

SuperSandro2000 commented on pkgs/applications/video/go2tv/default.nix:
```diff
@@ -0,0 +1,44 @@
+{ lib, buildGoModule, fetchFromGitHub, xorg, libglvnd, pkg-config, withGui ? true }:
+
+buildGoModule rec {
+  pname = if withGui then "go2tv" else "go2tv-lite";
```
```suggestion
  pname = "go2tv" + lib.optionalString (!withGui) "-lite";
```

SuperSandro2000 commented on pkgs/applications/video/go2tv/default.nix:
```diff
@@ -0,0 +1,44 @@
+
+  doCheck = false;
+
+  buildInputs = [
```
The withGui boolean must be used somewhere.

gdamjan replied:
I've improved it somewhat now, does it make more sense?

SuperSandro2000 replied:
Ah, that wasn't obvious to me. We don't need to add extra code for that but a nice short comment is enough.

gdamjan replied:
I could also make buildInputs an empty list for withGui=false perhaps?

(doesn't change the -lite binary and its dependencies though)

gdamjan commented:
> If we don't want to add it to a wrapper all the time there is none. Optional wrappers just to populate PATH are overkill.

maybe I should add a `withFfmpeg = false;` argument, so that if anyone wants to build manually they can do it?

SuperSandro2000 commented on pkgs/applications/video/go2tv/default.nix:
```diff
@@ -0,0 +1,48 @@
+    "-linkmode=external"
+  ];
+
+  subPackages = subPackage;
```
```suggestion
  # conditionally build with GUI or not
  subPackages = [ "cmd/${pname}" ];
```

SuperSandro2000 commented on pkgs/applications/video/go2tv/default.nix:
```diff
@@ -0,0 +1,48 @@
+{ lib, buildGoModule, fetchFromGitHub, xorg, libglvnd, pkg-config, withGui ? true }:
+
+let
+  cmdName = "go2tv" + lib.optionalString (!withGui) "-lite";
+  subPackage = "cmd/${cmdName}";
+in
+buildGoModule rec {
+  pname = cmdName;
```
```suggestion
buildGoModule rec {
  pname = "go2tv" + lib.optionalString (!withGui) "-lite";
```

SuperSandro2000 commented on maintainers/maintainer-list.nix:
```diff
@@ -4915,6 +4915,13 @@
     githubId = 37017396;
     name = "gbtb";
   };
+  gdamjan = {
```
Please put this into a separate commit with the message ``maintainers: add NAME`` and sort it alphabetically.

gdamjan committed:
maintainers: add gdamjan

gdamjan committed:
go2tv: init at 1.13.0

Cast media files to UPnP/DLNA Media Renderers and Smart TVs.
https://github.com/alexballas/go2tv

provides 2 packages:
- go2tv the GUI version (based on go-gl/glfw)
- go2tv-lite just a cli version

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

SuperSandro2000 commented on pkgs/applications/video/go2tv/default.nix:
```diff
@@ -0,0 +1,45 @@
+    xorg.libXext
+    xorg.libXxf86vm
+    libglvnd
+  ];
```
```suggestion
  ] ++ lib.optionals stdenv.isDarwin [ Carbon ];
```

SuperSandro2000 commented on pkgs/applications/video/go2tv/default.nix:
```diff
@@ -0,0 +1,45 @@
+{ lib, buildGoModule, fetchFromGitHub, xorg, libglvnd, pkg-config, withGui ? true }:
```
```suggestion
{ lib
, buildGoModule
, fetchFromGitHub
, Carbon
, xorg
, libglvnd
, pkg-config
, withGui ? true
}:
```

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2690,6 +2690,9 @@ with pkgs;
 
   gmnitohtml = callPackage ../applications/misc/gmnitohtml { };
 
+  go2tv = callPackage ../applications/video/go2tv { };
```
```suggestion
  go2tv = callPackage ../applications/video/go2tv {
    inherit (darwin.apple_sdk.frameworks) Carbon;
  };
```

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2690,6 +2690,9 @@ with pkgs;
   gmnitohtml = callPackage ../applications/misc/gmnitohtml { };
 
+  go2tv = callPackage ../applications/video/go2tv { };
+  go2tv-lite = callPackage ../applications/video/go2tv { withGui = false; };
```
```suggestion
  go2tv-lite = go2tv.override { withGui = false; };
```

SuperSandro2000 commented on pkgs/applications/video/go2tv/default.nix:
```diff
@@ -0,0 +1,45 @@
+    libglvnd
+  ];
+
+  nativeBuildInputs = [ pkg-config ];
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/applications/video/go2tv/default.nix:
```diff
@@ -0,0 +1,45 @@
+  vendorSha256 = "sha256-msXfXFWXyZeT6zrRPZkBV7PEyPqYkx+JlpTWUwgFavI=";
+
+  doCheck = false;
+
```
```suggestion

  nativeBuildInputs = [ pkg-config ];

```

SuperSandro2000 commented:
@ofborg build go2tv

SuperSandro2000 commented:
darwin now fails with 

```
# fyne.io/fyne/v2/internal/driver/glfw
menu_darwin.m:10:1: error: unknown type name 'NSControlStateValue'; did you mean 'NSCellStateValue'?
/nix/store/yydafvwas3hia0rl0f30mmhv2kh5lazx-apple-framework-AppKit/Library/Frameworks/AppKit.framework/Headers/NSCell.h:73:19: note: 'NSCellStateValue' declared here
menu_darwin.m:10:32: error: use of undeclared identifier 'NSControlStateValueOn'
menu_darwin.m:11:1: error: unknown type name 'NSControlStateValue'; did you mean 'NSCellStateValue'?
/nix/store/yydafvwas3hia0rl0f30mmhv2kh5lazx-apple-framework-AppKit/Library/Frameworks/AppKit.framework/Headers/NSCell.h:73:19: note: 'NSCellStateValue' declared here
menu_darwin.m:11:33: error: use of undeclared identifier 'NSControlStateValueOff'
```

gdamjan commented:
thanks @SuperSandro2000 for testing this, I don't have a way to test on darwin.

I see their Github Action for macos releases uses something called "fyne", checking it out now.

gdamjan commented:
maybe @alexballas has an idea why the build fails


only thing I've found is:
```
2017-12-22  Wenson Hsieh  <wenson_hsieh@apple.com>

        Unreviewed, try to fix the Sierra build after r226277.

        The macOS 10.12 SDK does not know about NSControlStateValue and some types of NSLevelIndicatorStyles, so these
        need to be declared in a separate section in AppKitCompatibilityDeclarations.h (rather than in the < macOS 10.12
        SDK #ifdef).

        * wtf/mac/AppKitCompatibilityDeclarations.h:```

SuperSandro2000 commented:
That sounds promising

SuperSandro2000 commented:
Lets try it with the newer sdk and if that does not work we mark it broken.

gdamjan commented:
```
app_darwin.m:6:9: fatal error: 'UserNotifications/UserNotifications.h' file not found
```

I've added PushKit too

SuperSandro2000 commented:
it builds on x84_64-darwin 🎉

gdamjan commented:
Thanks a lot :)
